compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1534,18 +1534,45 @@ impl<'a> Parser<'a> {
         Ok((ident, ItemKind::MacroDef(ast::MacroDef { body, macro_rules: false })))
     }
 
-    /// Is this unambiguously the start of a `macro_rules! foo` item definition?
-    fn is_macro_rules_item(&mut self) -> bool {
-        self.check_keyword(kw::MacroRules)
-            && self.look_ahead(1, |t| *t == token::Not)
-            && self.look_ahead(2, |t| t.is_ident())
+    /// Is this a possibly malformed start of a `macro_rules! foo` item definition?
+    ///
+    /// The first bool is whether it's a macro_rules item.
+    /// The second is whether the `!` after `macro_rules` is present
+    fn is_macro_rules_item(&mut self) -> (bool, bool) {
```

Comment:
nit: Seems like this code might be clearer if the return type were a struct with two boolean fields? E.g.:

```rust
struct IsMacroRulesItem {
    is_macro_rules: bool,
    has_bang_after_macro_rules: bool,
}
```

Reply:
Or return `Option<IsMacroRulesItem>` with `struct IsMacroRulesItem { has_bang: bool }`

Reply:
Went with 

```rust
enum IsMacroRulesItem {
    Yes { has_bang: bool },
    No,
}
```

we could bikeshed this forever, but I think this is pritty clear